lib/src/protocol_instance.rs

```diff
@@ -236,17 +239,25 @@ impl BlockMetaDataFork {
         }
     }
 
-    fn match_block_proposal(&self, other: &BlockProposedFork) -> bool {
+    fn match_block_proposal<'a>(
+        &'a self,
+        other: &'a BlockProposedFork,
+    ) -> (bool, Option<Box<dyn Display + 'a>>) {
         match (self, other) {
-            (Self::Hekla(a), BlockProposedFork::Hekla(b)) => a.abi_encode() == b.meta.abi_encode(),
-            (Self::Ontake(a), BlockProposedFork::Ontake(b)) => {
-                a.abi_encode() == b.meta.abi_encode()
-            }
-            (Self::Pacaya(a), BlockProposedFork::Pacaya(b)) => {
-                a.abi_encode() == b.meta.abi_encode()
-            }
-            (Self::None, BlockProposedFork::Nothing) => true,
-            _ => false,
+            (Self::Hekla(a), BlockProposedFork::Hekla(b)) => (
+                a.abi_encode() == b.meta.abi_encode(),
+                Some(Box::new(Comparison::new(a, &b.meta))),
+            ),
+            (Self::Ontake(a), BlockProposedFork::Ontake(b)) => (
+                a.abi_encode() == b.meta.abi_encode(),
+                Some(Box::new(Comparison::new(a, &b.meta))),
+            ),
+            (Self::Pacaya(a), BlockProposedFork::Pacaya(b)) => (
+                a.abi_encode() == b.meta.abi_encode(),
+                Some(Box::new(Comparison::new(a, &b.meta))),
+            ),
+            (Self::None, BlockProposedFork::Nothing) => (true, None),
+            _ => (false, None),
```

Comment:
Can we assert here directly, using `pretty_assertions::assert_eq`? It seems that it doesn't make sense to return outside and then assert outside.